crates/oro-node-semver/src/version_req.rs

```diff
@@ -15,25 +15,189 @@ use serde::ser::{Serialize, Serializer};
+            (Including(l), Unbounded, _, Including(r2)) => l <= r2,
+            (Including(l), Unbounded, _, Excluding(r2)) => l <= r2,
+            (Excluding(l), Unbounded, Unbounded, Including(r2)) => l < r2,
+            (Excluding(l), Unbounded, Unbounded, Excluding(r2)) => l <= r2,
+            (Excluding(l), Unbounded, _, Including(r2)) => l < r2,
+            (Unbounded, Excluding(r), _, Including(r2)) => r2 < r,
+            (Unbounded, Excluding(r), Excluding(l2), Unbounded) => l2 < r,
+            (Unbounded, Excluding(r), Including(l2), Unbounded) => l2 < r,
+            (Unbounded, Including(r), Including(l2), Including(_)) => l2 < r,
+            (Unbounded, Including(r), Excluding(l2), Unbounded) => l2 < r,
+            (Unbounded, Including(r), Including(l2), Unbounded) => l2 < r,
+            e => todo!("{:#?}", e),
+        }
+    }
+
+    fn intersect(&self, other: &Self) -> Option<Self> {
+        use Predicate::*;
+
+        let lower = match (&self.lower, &other.lower) {
+            (Unbounded, any) => any.clone(),
+            (any, Unbounded) => any.clone(),
+            (Including(v1), Including(v2)) => Including(std::cmp::max(v1, v2).clone()),
+            (Including(v1), Excluding(v2)) => {
+                if v2 < v1 {
+                    Including(v1.clone())
+                } else {
+                    Excluding(v2.clone())
+                }
+            }
+            (Excluding(v1), Excluding(v2)) => Excluding(std::cmp::max(v1, v2).clone()),
+            (Excluding(v1), Including(v2)) => {
+                if v2 < v1 {
+                    Excluding(v1.clone())
+                } else {
+                    Including(v2.clone())
+                }
+            }
+        };
+
+        let upper = match (&self.upper, &other.upper) {
+            (Unbounded, any) => any.clone(),
+            (any, Unbounded) => any.clone(),
+            (Including(v1), Including(v2)) => Including(std::cmp::min(v1, v2).clone()),
+            (Including(v1), Excluding(v2)) => {
+                if v1 < v2 {
+                    Including(v1.clone())
+                } else {
+                    Excluding(v2.clone())
+                }
+            }
+            (Excluding(v1), Excluding(v2)) => Excluding(std::cmp::min(v1, v2).clone()),
+            (Excluding(v1), Including(v2)) => {
+                if v1 <= v2 {
+                    Excluding(v1.clone())
+                } else {
+                    Including(v2.clone())
+                }
+            }
+        };
+
+        match (&lower, &upper) {
+            (Including(lower), Excluding(upper))
+            | (Excluding(lower), Excluding(upper))
+            | (Excluding(lower), Including(upper))
+                if upper <= lower =>
+            {
+                return None
+            }
+            (Including(lower), Including(upper)) if upper < lower => return None,
+            _ => {}
+        }
+
+        Some(Range { lower, upper })
+    }
```

Comment:
I don't know how to simplify this off the top of my head, but I did notice that `semver_constraint` uses comparisons to reduce these operations to something maybe simpler? I'm not convinced you'll be writing _that_ much less code, though. https://github.com/elba/semver_constraints/blob/master/src/lib.rs#L56

Reply:
Yeah, I'm trying this but stumbling a bit.